src/jdk.incubator.jextract/share/classes/jdk/internal/jextract/impl/OutputFactory.java

```diff
@@ -274,74 +245,59 @@ public Void visitFunction(Declaration.Function funcTree, Declaration parent) {
             return null;
         }
 
-        FunctionDescriptor descriptor = Type.descriptorFor(funcTree.type()).orElse(null);
-        if (descriptor == null) {
-            //abort
-            return null;
-        }
-
-        MemoryLayout unsupportedLayout = isUnsupported(descriptor);
-        if (unsupportedLayout != null) {
-            warn("skipping " + funcTree.name() + " because of unsupported type usage: " +
-                    UnsupportedLayouts.getUnsupportedTypeName(unsupportedLayout));
-            return null;
-        }
-
-        MethodType mtype = getMethodType(funcTree.type());
-        if (mtype == null) {
-            warn("skipping " + funcTree.name() + " because of unsupported type usage");
-            return null;
-        }
-
-        if (isVaList(descriptor)) {
-            MemoryLayout[] argLayouts = descriptor.argumentLayouts().toArray(new MemoryLayout[0]);
-            argLayouts[argLayouts.length - 1] = CLinker.C_VA_LIST;
-            descriptor = descriptor.returnLayout().isEmpty()?
-                    FunctionDescriptor.ofVoid(argLayouts) :
-                    FunctionDescriptor.of(descriptor.returnLayout().get(), argLayouts);
-            Class<?>[] argTypes = mtype.parameterArray();
-            argTypes[argLayouts.length - 1] = MemoryAddress.class;
-            mtype = MethodType.methodType(mtype.returnType(), argTypes);
-        }
-
         String mhName = Utils.javaSafeIdentifier(funcTree.name());
         //generate static wrapper for function
         List<String> paramNames = funcTree.parameters()
                                           .stream()
                                           .map(Declaration.Variable::name)
                                           .map(p -> !p.isEmpty() ? Utils.javaSafeIdentifier(p) : p)
                                           .collect(Collectors.toList());
-        int i = 0;
-        for (Declaration.Variable param : funcTree.parameters()) {
-            Type.Function f = getAsFunctionPointer(param.type());
-            if (f != null) {
-                String name = funcTree.name() + "$" + (param.name().isEmpty() ? "x" + i : param.name());
-                if (! generateFunctionalInterface(f, name)) {
-                    return null;
+
+        Optional<FunctionInfo> functionInfo = functionInfo(funcTree.type(), funcTree.name(), true,
+                (mtype, desc) -> FunctionInfo.ofFunction(mtype, desc, funcTree.type().varargs(), paramNames));
+
+        if (functionInfo.isPresent()) {
+            int i = 0;
+            for (Declaration.Variable param : funcTree.parameters()) {
+                Type.Function f = getAsFunctionPointer(param.type());
+                if (f != null) {
+                    String name = funcTree.name() + "$" + (param.name().isEmpty() ? "x" + i : param.name());
```

Comment:
This seems to let the name of the functional interface class depend on the parameter name? I thought we did _not_ want to do that?

Reply:
Ah, I see it's per-existing, sorry. I guess we can take care of it in another patch.